### ISSUE TITLE
Segmentation fault for larger minpack test

### ISSUE THREAD

**Issue by certik** [lfortran/lfortran#1624]
```diff
--- a/examples/example_hybrd.f90
+++ b/examples/example_hybrd.f90
@@ -10,7 +10,7 @@ program example_hybrd
 
     implicit none
 
-    integer,parameter :: n = 9
+    integer,parameter :: n = 900
     integer,parameter :: ldfjac = n
     integer,parameter :: lr = (n*(n+1))/2

```

This fails:
```console
$ lfortran examples/example_hybrd.f90
$ time ./a.out                       
zsh: segmentation fault  ./a.out
./a.out  0.00s user 0.00s system 49% cpu 0.006 total
```

**Comment by certik** [comment]
If I lower to `n=523`, then it runs and we can benchmark against GFortran: https://github.com/lfortran/lfortran/issues/1625.

**Comment by czgdp1807** [comment]
n = 900 means lr = 405450. We have `real(wp) :: r(lr)`. So, total of `8 * lr` bytes which means, `3.2` MB minimum memory required. Since arrays are allocated on stack by default, so depends if stack size is at least 3.2 MB.

**Comment by certik** [comment]
Yes, we can't be using stack for arrays. The macOS on Apple M1 has a small stack, so you run out of it quickly. I have 64GB memory, so a small 3.2 MB array must work well.

**Comment by czgdp1807** [comment]
I guess due to calling of `dogleg` subroutine in the following loop. We do `builder->CreateAlloca` for converting array descriptor to an argument array. So probably the stack space exhausts in one of these iterations. @certik If you can try adding a print statement which prints a counter to check before which call the code fails.

https://github.com/lfortran/lfortran/blob/e028e6a627110d27dbcf3eac4b2164f20fdde671/src/libasr/codegen/llvm_array_utils.cpp#L120-L126

```fortran
inner : do
    .
    .
    .
    call dogleg(n, r, Lr, Diag, Qtf, delta, Wa1, Wa2, Wa3)
    .
    .
    .
end do inner ! end of the inner loop.
```

**Comment by czgdp1807** [comment]
> I think every machine is different. On mine the maximum I can go to is 825.

Quoting from https://github.com/lfortran/lfortran/issues/1625#issuecomment-1529920405

**Comment by certik** [comment]
I think `llvm::Value* arg_struct = builder->CreateAlloca(arg_type, nullptr);` must be executed outside the loop at the top level subroutine level -- we have a mechanism for that, so we should use it.
